app/components/chat/use-file.ts

```diff
@@ -0,0 +1,155 @@
+import { v4 as uuidv4 } from "uuid";
+import {
+  DocumentFile,
+  DocumentFileType,
+  MessageAnnotation,
+  MessageAnnotationType,
+} from "@/cl/app/components/ui/chat";
+import { useClientConfig } from "@/cl/app/components/ui/chat/hooks/use-config";
+import { useBotStore } from "@/app/store/bot";
+
+const docMineTypeMap: Record<string, DocumentFileType> = {
+  "text/csv": "csv",
+  "application/pdf": "pdf",
+  "text/plain": "txt",
+  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
+    "docx",
+};
+
+export function useFile() {
+  // Custom useFile to get bot datasource
+  const botStore = useBotStore();
+  const bot = botStore.currentBot();
+
+  const { backend } = useClientConfig();
+  const [imageUrl, setImageUrl] = useState<string | null>(null);
+  const [files, setFiles] = useState<DocumentFile[]>([]);
+
+  const docEqual = (a: DocumentFile, b: DocumentFile) => {
+    if (a.id === b.id) return true;
+    if (a.filename === b.filename && a.filesize === b.filesize) return true;
+    return false;
+  };
+
+  const addDoc = (file: DocumentFile) => {
+    const existedFile = files.find((f) => docEqual(f, file));
+    if (!existedFile) {
+      setFiles((prev) => [...prev, file]);
+      return true;
+    }
+    return false;
+  };
+
+  const removeDoc = (file: DocumentFile) => {
+    setFiles((prev) => prev.filter((f) => f.id !== file.id));
+  };
+
+  const reset = () => {
+    imageUrl && setImageUrl(null);
+    files.length && setFiles([]);
+  };
+
+  const uploadContent = async (base64: string): Promise<string[]> => {
+    const uploadAPI = `${backend}/api/chat/upload`;
+    const response = await fetch(uploadAPI, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        base64,
+        datasource: bot.datasource,
+      }),
+    });
+    if (!response.ok) throw new Error("Failed to upload document.");
+    return await response.json();
+  };
+
+  const getAnnotations = () => {
+    const annotations: MessageAnnotation[] = [];
+    if (imageUrl) {
+      annotations.push({
+        type: MessageAnnotationType.IMAGE,
+        data: { url: imageUrl },
+      });
+    }
+    if (files.length > 0) {
+      annotations.push({
+        type: MessageAnnotationType.DOCUMENT_FILE,
+        data: { files },
+      });
+    }
+    return annotations as JSONValue[];
+  };
+
+  const readContent = async (input: {
+    file: File;
+    asUrl?: boolean;
+  }): Promise<string> => {
+    const { file, asUrl } = input;
+    const content = await new Promise<string>((resolve, reject) => {
+      const reader = new FileReader();
+      if (asUrl) {
+        reader.readAsDataURL(file);
+      } else {
+        reader.readAsText(file);
+      }
+      reader.onload = () => resolve(reader.result as string);
+      reader.onerror = (error) => reject(error);
+    });
+    return content;
+  };
+
+  const uploadFile = async (file: File) => {
```

Comment:
How about adding requestParams as a parameter here, an object of parameters that is passed to uploadContent. That way the caller of uploadFile can pass the dataSource and we can add this generic file to CL